src/maker/api.rs

```diff
@@ -672,10 +728,40 @@ pub(crate) fn check_for_idle_states(maker: Arc<Maker>) -> Result<(), MakerError>
 
 /// Broadcast Incoming and Outgoing Contract transactions & timelock contract after maturity or hashlock contract.
 /// Remove contract transactions from the wallet.
-pub(crate) fn recover_from_swap(maker: Arc<Maker>) -> Result<(), MakerError> {
-    // Broadcast all the incoming contracts.
+pub(crate) fn recover_from_swap(
+    maker: Arc<Maker>,
+    tracker_stream: Option<Arc<TcpStream>>,
+) -> Result<(), MakerError> {
     let (incomings, outgoings) = maker.wallet.read()?.find_unfinished_swapcoins();
     let is_hash_preimage_known = incomings.iter().any(|ic_sc| ic_sc.is_hash_preimage_known());
+    let timelock = outgoings
+        .first()
+        .map(|outgoing| outgoing.get_timelock())
+        .ok_or(MakerError::General(
+            "No timelock found on Outgoing Swapcoin",
+        ))??;
+
+    // Block wait time is varied between prod. and test builds.
+    let block_wait_time = if cfg!(feature = "integration-test") {
+        Duration::from_secs(10)
+    } else {
+        Duration::from_secs(10 * 60)
+    };
+
+    if let Some(tracker_stream) = &tracker_stream {
+        let time_till_timelock_expires =
+            Duration::from_secs(timelock as u64 * block_wait_time.as_secs());
+        let start_time = Instant::now();
+
+        while start_time.elapsed() < time_till_timelock_expires {
+            if incomings.iter().any(|ic_sc| ic_sc.is_hash_preimage_known()) {
+                break;
+            }
+
+            check_for_tracker_watch_response(&maker, &mut tracker_stream.clone())?;
+            sleep(Duration::from_secs(10));
```

Comment:
You also need to break from this loop the moment hashpreimage is known. Or else it will keep looping until timelock expires, even if hash preimage is known.